### ISSUE TITLE
[FR] Remove additional fetch calls during recurseRelators calls.

### ISSUE THREAD

**Issue by jaridmargolin** [mathematic-inc/ts-japi#24]
## Is your feature request related to a problem? <sup>*</sup>

The `Relator` fetch method is excessively called during `recurseRelators` due to the algorithm neglecting to remove finished work from queue.

As an example:

**Code:**
```ts
  it("Should bypass recurse cycles after data fetched", async () => {
    type A = { id: string; prop: string }
    type B = { id: string; prop: string }
    let fetchCounter = 0

    const SerializerB = new Serializer<B>("b")
    const AtoBRelator = new Relator<A, B>(async (data) => {
      fetchCounter++
      return { id: "1", prop: "b" }
    }, SerializerB)

    const SerializerA = new Serializer<A>("a", {
      relators: { b: AtoBRelator },
    })

    const serialized = await SerializerA.serialize(
      { id: "1", prop: "a" },
      { depth: 10 }
    )

    console.log(fetchCounter)
  })
```

**Output:**
```
  console.log
    11
```

## Describe the solution you'd like<sup>*</sup>

If the work is removed from the queue after resolving, the algo can exit earlier :) I will submit a followup PR.

## Additional context

I believe there may be further optimizations possible to reduce duplicate fetch calls during serialization, but this is an easy first pass.

**Comment by jrandolf** [comment]
The excessive fetch calls from the original algorithm have been fixed, however I do want to remove the additional call.

From what I investigated, we can carry the depth around to other serializers and cache fetches from relationships based on the depth.

**Comment by jaridmargolin** [comment]
I'd really like to see the additional call removed as well. I'm prying around but I'm not seeing a simple patch to fix this.

First, are you itching to take a crack at this? I'd hate to spend my time working on this, only to find out we were both working in parallel. If you're busy elsewhere, I don't mind trying to take a stab at this, but I think I may see things differently than you (which I am sure is due to my limited exposure). 

...here are my observations so far:

The fetch is called in two distinct places (perhaps there is more but these are the places getting hit in the code paths I see).
1. The Serializer, inside of the `createResource` method: `relationships[name] = await relator.getRelationship(data);`
2. Inside of the `recurseRelators` method.

As far as I understand, a cache cannot be added to the `relator` because that relator get's used across multiple serializations. The cache would need to be local to the `serialize` method being called (or instances that are scoped to the serialize method, such as the `Resource`).

My initial instinct would be to remove data fetching from the `recurseRelators` method, and instead rely on the fetch happening when creating the `Resource` relationships. The recurse method would then just pull the cached value from some sort of relationships cache living on the `Resource`.

**Comment by jrandolf** [comment]
Since we do not use this library in our stack anymore, we are looking to just maintain this, so you can take a stab at it.

Also, the cache I meant was local, not global, so our solutions are essentially the same (remove the tossing of the depth arg).

It might be worth noting that although this may save some calls, practically speaking, the solution is still far from optimal.

Optimally, we should have a separate relationship fetcher (something that fetches table joins for example) and replace relationship data fetches with these ones. In this way, we can fetch the relationship and use the relationship to get extra data (if needed), rather than fetching everything just for a relationship (which may or may not be included as an included resource). This solution is the most natural solution, but also breaks the API.

(Side Note: there was a misunderstanding early in development about the spec which is the reason why this optimal solution wasn't implemented)

**Comment by jaridmargolin** [comment]
Ok. I have spent some time trying to plan how exactly this could be implemented and I have exactly one viable solution, but I wanted to discuss it prior to implementing. Jumping back to my message above, the reason duplicate calls are seen is due to the relators being fetched in two places (in order):

1. When populating the` Resource` `Relationship`s during the `createResource` call.
2. During the `recurseRelators` step which is used to populate `document.includes`.

The solution, as I see it, is to cache the result found during the `createResource` call as it occurs first.

This is where I ran into the first challenge. `relator.getRelatedData` (fetch) is called within the body of `relator.getRelationship` method, which returns an instance of the `Relationship` model. So the tricky part is, if we can't change the publicly exposed API of the relator (the return value), how can we get the cached result up to our outer `createResource` method? The best solution I can see, is to pass a second argument which is capable of capturing the value.

**relator**
```ts
type CacheDataFn = <Data extends RelatedType | RelatedType[] | nullish>(data: Data) => Data
getRelationship(data: PrimaryType, cacheData: CacheDataFn) => {
  ...
  const relatedData = cacheData(await this.getRelatedData(data));
  ...
}
```

**serializer**
```ts
createResource() {
   ...
   relationships[name] = await relator.getRelationship(data, (data) => (resourceRelationshipsCache[name] = data));
   ...
}
```

And in similar nature, the `createResource` method returns a `Resource` model, so I'd need to thread the `resourceRelationshipsCache` in as some type of argument to the `createResource` method so that I can have access to the cache in the outer `serialize` method.

This cache could then be passed in and utilized in the `recurseRelators` util. Modifications would also have to be made within the `recurseRelators` internals to ensure it cached the relationships for each resource it creates.

It's not the prettiest of solutions, but it's the only solution I could devise that didn't semantically break the system (patching non-enumerable properties on to the model), and didn't cause any breaking changes to the publicly exposed APIs.
